src/xenialdan/MagicWE2/API.php

```diff
@@ -136,27 +137,32 @@ public static function fill(Selection $selection, ?Session $session, $newblocks
         $changed = 0;
         $time = microtime(TRUE);
         try {
-            $blocks = [];
+            $preBlocks = $selection->getBlocks($flags);
+            $postBlocks = [];
             /** @var Block $block */
-            foreach ($selection->getBlocks($flags) as $block) {
-                $level = $selection->getLevel() ?? (!is_null($session) ? $session->getPlayer()->getLevel() : $block->getLevel());
-                if ($block->y >= Level::Y_MAX || $block->y < 0) continue;
-                if (API::hasFlag($flags, API::FLAG_HOLLOW) && ($block->x > $selection->getMinVec3()->getX() && $block->x < $selection->getMaxVec3()->getX()) && ($block->y > $selection->getMinVec3()->getY() && $block->y < $selection->getMaxVec3()->getY()) && ($block->z > $selection->getMinVec3()->getZ() && $block->z < $selection->getMaxVec3()->getZ())) continue;
-                $newblock = $newblocks[array_rand($newblocks, 1)];
-                $newblock->position($block->asPosition());
-                if (API::hasFlag($flags, API::FLAG_KEEP_BLOCKS)) {
-                    if ($level->getBlock($block)->getId() !== Block::AIR) continue;
-                }
-                #if (API::hasFlag($flags, API::FLAG_KEEP_AIR)){
-                #	if ($level->getBlock($block)->getId() === Block::AIR) continue;
-                #}
-                if ($level->setBlock($block, $newblock, false, false)) {
-                    $blocks[] = $block;
-                    $changed++;
+            foreach ($preBlocks as $block) {
+                $new = clone $newblocks[array_rand($newblocks, 1)];
+                $new->position($block->asPosition());
+                $postBlocks[] = $new;
+            }
+            //Event call
+            $event = new MWEEditEvent(Loader::getInstance(), $preBlocks, $postBlocks, $session);
+            Loader::getInstance()->getServer()->getPluginManager()->callEvent($event);
```

Comment:
Just a note in passing: Once 3.4 is released, if you drop support for everything < 3.4 you'll be able to write `$event->call()` instead of this boilerplate.

Reply:
Sure, knew about that already, and the change is already done locally